### PR TITLE
Desktop: Use system keychain for storing passwords

### LINKED ISSUE #2773: Save passwords to system keychain
The current version of Joplin saves the sync passwords to the database. Normally, this is secure as long as the system itself is secure.

However for improved security, the passwords should be saved to the system keychain. This can be done in a cross platform way using https://github.com/atom/node-keytar

The difficulty will be too read and write these passwords in a transparent way from the Setting class.

### PR DESCRIPTION
**Pull request by jyuvaraj03**
Closes #2773 

This PR saves the passwords to the system keychain using the [keytar](https://github.com/atom/node-keytar) package.

The passwords were already marked with `secure: true` in `this.metadata_`. So, any property with `secure: true` in the metadata (sync.5.password, sync.6.password, encryption.passwordCache) was saved to the system keychain rather than the database. 

I didn't write automated tests for this, but I have manually tested if the passwords were saved and retrieved each time the app is launched. And everything seems to be working exactly as it was before.

### CODE REVIEW

**Comment by jyuvaraj03** [comment]
@PackElend please label me

File: ReactNativeClient/lib/models/Setting.js [hunk 1]
```diff
@@ -661,7 +666,24 @@ class Setting extends BaseModel {
 
 				this.cache_.push(c);
 			}
+			if (this.constants_.appType !== 'mobile') this.loadSecureKeys();
+			this.dispatchUpdateAll();
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js]: Please don't check for app type here. Instead use lib/shim.js to abstract away differences between platforms. For example create a function in the shim that returns all the secure credentials. On Node/Electron, make it use keytar. On React Native, since it's not implemented, make it return an empty array.

Make sure that the logic means that it works both on RN and Node.

File: ReactNativeClient/lib/models/Setting.js [hunk 2]
```diff
@@ -918,7 +940,11 @@ class Setting extends BaseModel {
 		for (let i = 0; i < this.cache_.length; i++) {
 			const s = Object.assign({}, this.cache_[i]);
 			s.value = this.valueToString(s.key, s.value);
-			queries.push(Database.insertQuery(this.tableName(), s));
+			if (this.constants_.appType !== 'mobile' && this.isSecureKey(s.key)) {
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js]: Again no device check in Setting.js. Use the shim

File: ReactNativeClient/lib/models/Setting.js (hunk 1 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/models/Setting.js]: I wasn't aware of the shim.js file. I'll make use of it now!

File: ReactNativeClient/lib/models/Setting.js (hunk 2 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/models/Setting.js]: Sure!

**Comment by jyuvaraj03** [comment]
> We'll also need this:
> * Unit tests

I'll write the unit tests now. 

> * Also please make sure this feature is not enabled in the portable versions, as it shouldn't use any keychain in this case.

Yeah, I'll do that. But, I'm curious as to what portable versions mean? Does it mean the mobile version?

> * Also what happens if the system doesn't have a keychain?

If I'm right, macOS and Windows have the keychain inbuilt and do not need separate installations. But, the keytar library currently uses `libsecret` in its implementation for Linux. `libsecret` needs a separate installation. I'm confused on how to do this. Because installing `libsecret` will need `sudo` access. So, it can't be installed by our install scripts, right? Should we ask the users to manually install `libsecret` mandatorily or should this whole feature be optional?

**Comment by laurent22** [comment]
> But, I'm curious as to what portable versions mean? 

The JoplinPortable.exe version that's in each release.

> So, it can't be installed by our install scripts, right? 

No, your code should detect if a keychain is available or not. If it's not, it should default to save to the database. Joplin, and in particular the CLI  app, can be used on many different systems, including beOs, etc. so don't assume a keychain will be available.

In fact please also double-check that your code works in cli.

**Comment by jyuvaraj03** [comment]
I have written the unit tests and made sure the implementation does not depend on having a keychain installed. If the keychain is not present, the passwords are saved to the database itself. 

Please review and let me know if any changes are required.

File: ReactNativeClient/lib/shim-init-node.js [hunk 3]
```diff
@@ -12,6 +12,14 @@ const { _ } = require('lib/locale.js');
...
 const https = require('https');
 const toRelative = require('relative');
+const { Database } = require('lib/database.js');
+
+let keytar;
+try {
+	keytar = require('keytar');
+} catch (err) {
+	keytar = null;
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js]: `keytar` is installed and always available so no need for a try/catch block

File: CliClient/tests/models_Setting.js [hunk 4]
```diff
@@ -31,4 +31,19 @@ describe('models_Setting', function() {
 		expect('username' in output).toBe(false);
 	}));
 
+	it('should save values', asyncTest(async () => {
```
Comment (laurent22) [inline CliClient/tests/models_Setting.js]: That's not specific to the keychain feature

File: CliClient/tests/models_Setting.js [hunk 5]
```diff
@@ -31,4 +31,19 @@ describe('models_Setting', function() {
...
+		await setupDatabase();
```
Comment (laurent22) [inline CliClient/tests/models_Setting.js]: should be in beforeEach

File: CliClient/tests/models_Setting.js [hunk 6]
```diff
@@ -31,4 +31,19 @@ describe('models_Setting', function() {
...
+
+		let key = 'sync.5.username';
+		let value = 'testUser';
+		await Setting.setValue(key, value);
```
Comment (laurent22) [inline CliClient/tests/models_Setting.js]: setValue is not async

File: CliClient/tests/models_Setting.js [hunk 7]
```diff
@@ -31,4 +31,19 @@ describe('models_Setting', function() {
...
+		let savedValue = Setting.value(key);
+		expect(savedValue).toBe(value);
+
+		key = 'sync.5.password';
+		value = 'Test$123';
+		await Setting.setValue(key, value);
+		savedValue = Setting.value(key);
+		expect(savedValue).toBe(value);
```
Comment (laurent22) [inline CliClient/tests/models_Setting.js]: I don't think this tests anything related to this feature, because the test env may or may not have a keychain.

File: ReactNativeClient/lib/shim-init-node.js [hunk 8]
```diff
@@ -412,6 +420,49 @@ function shimInit() {
...
+			for (let i = 0; i < items.length; i++) {
+				const c = items[i];
+
+				c.value = Setting.formatValue(c.key, c.value);
+				c.value = Setting.filterValue(c.key, c.value);
+
+				Setting.cache_.push(c);
+			}
+			Setting.dispatchUpdateAll();
+		}
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js]: loadSecureItems should only do the absolute minimum related to keychain - it should return an array with the relevant credentials and that's it. There's shouldn't be Setting related code, let alone dispatching actions, in here. That also means refactoring the call in Setting.js

File: ReactNativeClient/lib/shim-init-node.js [hunk 9]
```diff
@@ -412,6 +420,49 @@ function shimInit() {
...
+				keytar.setPassword(Setting.constants_.appId, items[i].key, items[i].value);
+			}
+		} else {
+			const queries = [];
+			for (let i = 0; i < items.length; i++) {
+				queries.push(Database.insertQuery(Setting.tableName(), items[i]));
+			}
+			await Setting.db().transactionExecBatch(queries);
+		}
+	};
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js]: Same comment as above. That function should save a secure item and that's it.

File: ReactNativeClient/lib/shim-init-react.js [hunk 10]
```diff
@@ -191,6 +192,26 @@ function shimInit() {
...
+		Setting.dispatchUpdateAll();
+	};
+
+	shim.saveSecureItems = async (Setting, items) => {
+		const queries = [];
+		for (let i = 0; i < items.length; i++) {
+			queries.push(Database.insertQuery(Setting.tableName(), items[i]));
+		}
+		await Setting.db().transactionExecBatch(queries);
+	};
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-react.js]: If you refactor the functions as described above you won't need all this duplicated code.

File: ReactNativeClient/lib/shim-init-node.js (hunk 3 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js]: The reason I used a try/catch block here is because when the system doesn't have a keychain, keytar threw an error, even when keytar itself was installed.

File: CliClient/tests/models_Setting.js (hunk 7 above)
Comment (jyuvaraj03) [reply CliClient/tests/models_Setting.js]: I'm having trouble writing a test for this. I am not sure what exactly I should be testing. Can you tell what I should be testing for? Should I be directly retrieving from the system keychain (if present) and test? I need some help here!

File: ReactNativeClient/lib/shim-init-node.js (hunk 8 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js]: Yes, that makes sense. I'll refactor the code as you suggested.

File: ReactNativeClient/lib/shim-init-react.js (hunk 10 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-react.js]: You're right. I'll refactor the code to avoid duplication.

File: CliClient/tests/models_Setting.js (hunk 6 above)
Comment (jyuvaraj03) [reply CliClient/tests/models_Setting.js]: Sorry, that was a mistake. I'll rewrite the tests.

**Comment by jyuvaraj03** [comment]
I have rewritten the code as you suggested. But, I'm stuck on how to write unit tests for this feature. What should I be testing? The `shim.loadSecureItems()` and `shim.saveSecureItem()` functions? Please suggest what I should be doing. Thanks!

File: ReactNativeClient/lib/shim-init-node.js (hunk 3 above)
Comment (laurent22) [reply ReactNativeClient/lib/shim-init-node.js:21]: Really? Ok that's not great, but in that case your implementation indeed makes sense. Thanks for looking into it.

Comment (laurent22) [reply ReactNativeClient/lib/shim-init-node.js:21]: Actually please could you add a comment then about your try/catch block to explain why it's there?

File: ReactNativeClient/lib/models/Setting.js [hunk 11]
```diff
@@ -652,17 +656,32 @@ class Setting extends BaseModel {
...
+				// secureItems is an array of { account: 'foo', password: 'bar' }
+				// but each item should be in the form of { key: 'foo', value: 'bar' }
+				// https://stackoverflow.com/a/50101979
+				for (const item of secureItems) {
+					delete Object.assign(item, { ['key']: item['account'] })['account'];
+					delete Object.assign(item, { ['value']: item['password'] })['password'];
+				}
+				pushItemsToCache(secureItems);
+				this.dispatchUpdateAll();
+			});
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js:684]: The logic is not right here because the `load()` function will exit before your code here has run, which could cause race conditions. I think you should change this to `return shim.loadSecureItems(...` although ideally all this method should be refactored with async/await.

File: ReactNativeClient/lib/models/Setting.js [hunk 12]
```diff
@@ -918,6 +937,11 @@ class Setting extends BaseModel {
 		for (let i = 0; i < this.cache_.length; i++) {
 			const s = Object.assign({}, this.cache_[i]);
 			s.value = this.valueToString(s.key, s.value);
+
+			if (this.isSecureKey(s.key)) {
+				if (shim.saveSecureItem(this.constants_.appId, s)) continue;
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js:942]: Should be `await shim.saveSecureItem(....`

File: ReactNativeClient/lib/shim-init-node.js [hunk 13]
```diff
@@ -412,6 +419,24 @@ function shimInit() {
...
+		}
+		return Promise.resolve([]);
+	};
+
+	shim.saveSecureItem = (appId, item) => {
+		if (shim.isKeytarAvailable()) {
+			return keytar.setPassword(appId, item.key, item.value);
+		}
+		return false;
+	};
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js:439]: Please make these methods async. Keytar might be synchronous, but it's likely that for example a React Native lib would be async, so it needs to work with both.

File: ReactNativeClient/lib/shim-init-node.js [hunk 14]
```diff
@@ -412,6 +419,24 @@ function shimInit() {
 	shim.pathRelativeToCwd = (path) => {
 		return toRelative(process.cwd(), path);
 	};
+
+	shim.isKeytarAvailable = () => {
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js:423]: As this is only used by your code, please remove it from the shim and keep as a private helper method.

File: ReactNativeClient/lib/shim.js [hunk 15]
```diff
@@ -220,4 +220,16 @@ shim.pathRelativeToCwd = (path) => {
 	throw new Error('Not implemented');
 };
 
+shim.isKeytarAvailable = () => {
+	return false;
+};
```
Comment (laurent22) [inline ReactNativeClient/lib/shim.js:225]: Can be removed.

File: ReactNativeClient/lib/shim.js [hunk 16]
```diff
@@ -220,4 +220,16 @@ shim.pathRelativeToCwd = (path) => {
...
+
+shim.loadSecureItems = (appId) => {
+	return Promise.resolve([]);
+};
+
+shim.saveSecureItem = (appId, item) => {
+	return false;
+};
```
Comment (laurent22) [inline ReactNativeClient/lib/shim.js:233]: Please make async

File: ReactNativeClient/lib/models/Setting.js [hunk 17]
```diff
@@ -652,17 +656,32 @@ class Setting extends BaseModel {
...
+				}
+			};
 
-			this.dispatchUpdateAll();
+			pushItemsToCache(rows);
+
+			shim.loadSecureItems(this.constants_.appId).then(secureItems => {
+				// secureItems is an array of { account: 'foo', password: 'bar' }
+				// but each item should be in the form of { key: 'foo', value: 'bar' }
+				// https://stackoverflow.com/a/50101979
```
Comment (laurent22) [inline ReactNativeClient/lib/models/Setting.js:677]: Could you abstract this away and make `shim.loadSecureItems` return an array of `{ key: 'foo', value: 'bar' }`?

**Comment by laurent22** [comment]
> I have rewritten the code as you suggested. But, I'm stuck on how to write unit tests for this feature. What should I be testing? The `shim.loadSecureItems()` and `shim.saveSecureItem()` functions? Please suggest what I should be doing. Thanks!

Now that you've converted these methods to smaller one it should be easier to unit test. I believe you would need to mock the methods to return whatever you need for testing. For example in a test unit, you could have `shim.loadSecureItems = async() => [{ key: 'foo', value: 'bar' }, { key: 'foo2', value: 'bar2' }]`. Then you call `Setting.load()` and verify that your keys/passwords are in there.

Similar idea for saving settings, you could mock it that way:

```
let myKeys = []
shim.saveSecureItem = (appId, item) => {
    myKeys.push(item);
}
```

Then call `Setting.setValue(......`  and verify that the keys get into your `myKeys` array.

File: ReactNativeClient/lib/shim-init-node.js (hunk 3 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js:21]: Yes, sure.

File: ReactNativeClient/lib/models/Setting.js (hunk 17 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/models/Setting.js:677]: Sure, will do it.

File: ReactNativeClient/lib/shim.js (hunk 16 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim.js:233]: I'll make both of the methods as async.

File: ReactNativeClient/lib/models/Setting.js (hunk 11 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/models/Setting.js:684]: I hadn't considered that. I will refactor them with async/await.

**Comment by jyuvaraj03** [comment]
> For example in a test unit, you could have shim.loadSecureItems = async() => [{ key: 'foo', value: 'bar' }, { key: 'foo2', value: 'bar2' }]. Then you call Setting.load() and verify that your keys/passwords are in there.

This made it very clear for me on how to test. Thanks for explaining!

File: ReactNativeClient/lib/shim-init-node.js (hunk 14 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js:423]: By private helper method, you mean a separate file -- something like `keytar-utils.js` in which the function will reside. Am I right?

Comment (laurent22) [reply ReactNativeClient/lib/shim-init-node.js:423]: No, it can be a simple function in shim-init-node. Just `function isKeytarAvailable() { .... }`

Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js:423]: Oh okay. Thanks!

**Comment by jyuvaraj03** [comment]
Just a reminder. I have implemented all the requested changes and the PR is ready for review!

File: CliClient/tests/models_Setting.js [hunk 18]
```diff
@@ -31,4 +34,51 @@ describe('models_Setting', function() {
...
 	}));
 
+	it('should save secure items', asyncTest(async () => {
+		const originalScheduleSave = Setting.scheduleSave;
+		const originalSaveSecureItem = shim.saveSecureItem;
+
+		Setting.scheduleSave = async () => {
+			Setting.saveTimeoutId_ = 123;
+			await Setting.saveAll();
+		};
```
Comment (laurent22) [inline CliClient/tests/models_Setting.js:44]: Why did you need to override this method?

File: ReactNativeClient/lib/shim-init-node.js [hunk 19]
```diff
@@ -445,6 +455,33 @@ function shimInit() {
...
+		return shim.isElectron() && !shim.isPortable() && keytar;
+	}
+
+	shim.loadSecureItems = async (appId) => {
+		if (isKeytarAvailable()) {
+			const secureItems = await keytar.findCredentials(appId);
+			// secureItems is an array of { account: 'foo', password: 'bar' }
+			// but it should be an array of { key: 'foo', value: 'bar' }
+			// https://stackoverflow.com/a/50101979
+			for (const item of secureItems) {
```
Comment (laurent22) [inline ReactNativeClient/lib/shim-init-node.js:469]: No mutation, so please create a new array `newSecureItems` and then `newSecureItems.push({ key: .... value: ....})`

**Comment by jyuvaraj03** [comment]
> By the way, how does it handle existing secure items. For example, if I have a passwordCache in my setting table, does it move it to the keychain? Does it override my setting with empty data?

Yes, the existing secure items are moved to the keychain (if one exists). This happens automatically when the user opens this version for the first time. 

If the user has any secure items from the previous version, it is stored in the database. So, when the user boots up the app, the secure items are loaded to the app cache by the `load()` method. Now, when `scheduleSave()` calls the `saveAll()` method, the secure items are stored to the system keychain and the existing secure items in the database are deleted during this action. All of these happen with the existing code itself, so I didn't have to write code to do it explicitly.

File: CliClient/tests/models_Setting.js (hunk 18 above)
Comment (jyuvaraj03) [reply CliClient/tests/models_Setting.js:44]: I had to override this method so that when I called `Setting.setValue(key, value)`, the key-value pair is immediately saved. Otherwise, the save is performed asynchronously and I couldn't check if the values were correctly saved.

File: ReactNativeClient/lib/shim-init-node.js (hunk 19 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js:468]: I've done this. But, why is mutation not preferred? I'm trying to learn :grimacing:

File: CliClient/tests/models_Setting.js (hunk 18 above)
Comment (laurent22) [reply CliClient/tests/models_Setting.js:44]: Right, that makes sense. However you can simply set the value and then call `await Setting.saveAll()` just after, and they will be saved them immediately. Could you make this change please?

File: ReactNativeClient/lib/shim-init-node.js (hunk 19 above)
Comment (laurent22) [reply ReactNativeClient/lib/shim-init-node.js:468]: In this particular case, we don't really know how `findCredentials` works internally - probably it returns a new array every time you call it. Probably. But it could also for example cache the data and return the same array every time. Then if you modify that array, you are changing data used by keytar and could break something.

I don't think it would actually happen, but to be safe it always better not to mutate data, especially data you didn't create yourself. It's mostly defensive programming, because object ownership it not well defined in JavaScript.

**Comment by laurent22** [comment]
> If the user has any secure items from the previous version, it is stored in the database. So, when the user boots up the app, the secure items are loaded to the app cache by the `load()` method. Now, when `scheduleSave()` calls the `saveAll()` method, the secure items are stored to the system keychain and the existing secure items in the database are deleted during this action. All of these happen with the existing code itself, so I didn't have to write code to do it explicitly.

That makes perfect sense, thanks for clarifying.

**Comment by laurent22** [comment]
Ok I've merged in a "keytar" branch for now as I want to make a few tweaks. In particular the CI config needs to be updated for Linux and the package should be removed from the mobile client (since it's not compatible with it). Otherwise, it works well, thanks for your efforts @jyuvaraj03!

File: ReactNativeClient/lib/shim-init-node.js (hunk 19 above)
Comment (jyuvaraj03) [reply ReactNativeClient/lib/shim-init-node.js:468]: Thanks for explaining!

**Comment by PackElend** [comment]
@laurent22, will you reference the necessary tweaks to this PR?
